File: clients/android/sample/src/androidTest/kotlin/dev/arizona/sample/ListE2ETest.kt
package dev.arizona.sample

import androidx.compose.ui.test.junit4.createAndroidComposeRule
import androidx.compose.ui.test.onNodeWithText
import androidx.compose.ui.test.performClick
import kotlinx.serialization.json.addJsonObject
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import kotlinx.serialization.json.putJsonArray
import org.junit.Rule
import org.junit.Test

/**
 * On-device (emulator) e2e for the keyed list -- the Kotlin-renderer analogue of
 * e2e/native/list.spec.js. Navigate to /native/list, then drive the stream ops
 * directly via the client (the list view has no buttons), exercising
 * OP_INSERT / OP_REMOVE / OP_MOVE / OP_ITEM_PATCH / OP_UPDATE on the renderer.
 * Payloads carry real JSON types (numeric `pos`, an array for `reset`), which
 * the Map<String,String> signature couldn't express.
 *
 * Presence/absence/text assertions verify the diff applied; exact positional
 * order after a move stays in the JS spec, where tree introspection is clean.
 *
 * Requires the Arizona test server reachable at localhost:4040 (adb reverse
 * tcp:4040 tcp:4040). Run with: ./gradlew :sample:connectedCheck
 */
class ListE2ETest {
    @get:Rule
    val rule = createAndroidComposeRule<MainActivity>()

    @Test
    fun appliesStreamOpsOverTheWire() {
        // Launcher shows the menu; navigate to the list (initial 3 keyed items).
        rule.waitForText("List")
        rule.onNodeWithText("List").performClick()
        rule.waitForText("One")
        rule.waitForText("Two")
        rule.waitForText("Three")

        val client = rule.activity.client

        // OP_INSERT (append).
        client.pushEvent("add", buildJsonObject { put("id", "9"); put("text", "Nine") })
        rule.waitForText("Nine")

        // OP_REMOVE.
        client.pushEvent("remove", buildJsonObject { put("id", "2") })
        rule.waitForNoText("Two")

        // OP_MOVE: numeric pos -- '1' to index 2. The item stays present (exact
        // order is asserted in the JS spec).
        client.pushEvent("move", buildJsonObject { put("id", "1"); put("pos", 2) })
        rule.waitForText("One")

        // OP_ITEM_PATCH: update item 9's text in place.
        client.pushEvent("update", buildJsonObject { put("id", "9"); put("text", "Updated") })
        rule.waitForText("Updated")
        rule.waitForNoText("Nine")

        // OP_UPDATE (full re-render via reset) -- an array-of-objects payload.
        client.pushEvent(
            "reset",
            buildJsonObject {
                putJsonArray("items") {
                    addJsonObject { put("id", "a"); put("text", "A") }
                    addJsonObject { put("id", "b"); put("text", "B") }
                }
            },
        )
        rule.waitForText("A")
        rule.waitForText("B")
        rule.waitForNoText("Three")
    }
}

File: clients/ios/Sample/UITests/ListE2ETest.swift
import XCTest

/// On-device keyed-list e2e -- the analogue of the Android ListE2ETest. The list
/// view has no buttons, so the stream ops are driven via the `-uitest` debug
/// toolbar (each button calls client.pushEvent with the same payloads the Android
/// test sends), exercising OP_INSERT / OP_REMOVE / OP_MOVE / OP_ITEM_PATCH /
/// OP_UPDATE over a real socket. XCUITest is out-of-process and can't call the
/// client directly, so the toolbar is the stand-in.
final class ListE2ETest: XCTestCase {
    private var app: XCUIApplication!

    override func setUpWithError() throws {
        continueAfterFailure = false
        app = launchSample()
    }

    func testAppliesStreamOpsOverTheWire() {
        tapButton(app, "List")
        waitForText(app, "One")
        waitForText(app, "Two")
        waitForText(app, "Three")

        // OP_INSERT (append) -- add {id: 9, text: Nine}.
        tapButton(app, "dbg-add")
        waitForText(app, "Nine")

        // OP_REMOVE -- remove {id: 2}.
        tapButton(app, "dbg-remove")
        waitForNoText(app, "Two")

        // OP_MOVE -- numeric pos, move {id: 1} to index 2 (item stays present).
        tapButton(app, "dbg-move")
        waitForText(app, "One")

        // OP_ITEM_PATCH -- update item 9's text in place.
        tapButton(app, "dbg-update")
        waitForText(app, "Updated")
        waitForNoText(app, "Nine")

        // OP_UPDATE (full re-render via reset) -- an array-of-objects payload.
        tapButton(app, "dbg-reset")
        waitForText(app, "A")
        waitForText(app, "B")
        waitForNoText(app, "Three")
    }
}
